### PR TITLE
Paged response for a list of wells

### PR DESCRIPTION
**Pull request by mgcam**

### CODE REVIEW

File: lang_qc/endpoints/pacbio_well.py [hunk 1]
```diff
@@ -78,17 +78,26 @@
...
+    page_number:PositiveInt,
     qcdb_session: Session = Depends(get_qc_db),
     mlwh_session: Session = Depends(get_mlwh_db),
 ):
 
-    wells, states = grab_wells_with_status(qc_status, qcdb_session, mlwh_session, weeks)
-    return pack_wells_and_states(wells, states)
+    pager = PagedRequest(page_size = page_size, page_number=page_number)
+    wells, states = grab_wells_with_status(qc_status, qcdb_session, mlwh_session)
+    pbwells = pack_wells_and_states(wells, states)
```
Comment (nerdstrike) [inline lang_qc/endpoints/pacbio_well.py:93]: I think I would be happier if `pack_wells_and_states` went away and was replaced by something more coherent. I think it's hinding and doing too much that is not immediately obvious

**Review by nerdstrike — COMMENTED** [review]
The result that will be sent over the wire should be perfectly adequate, and the heritable pager model is a good idea. I'm not so fond of the implementation!

**Comment by mgcam** [comment]
> The result that will be sent over the wire should be perfectly adequate, and the heritable pager model is a good idea. I'm not so fond of the implementation!

As discussed, I am not fond of teh implementation either. Will do for now, but shoudl be revisited when we have product ids for wells in mlwh.

File: lang_qc/endpoints/pacbio_well.py [hunk 2]
```diff
@@ -78,17 +78,26 @@
...
+    page_size: PositiveInt,
+    page_number:PositiveInt,
     qcdb_session: Session = Depends(get_qc_db),
     mlwh_session: Session = Depends(get_mlwh_db),
 ):
 
-    wells, states = grab_wells_with_status(qc_status, qcdb_session, mlwh_session, weeks)
-    return pack_wells_and_states(wells, states)
+    pager = PagedRequest(page_size = page_size, page_number=page_number)
+    wells, states = grab_wells_with_status(qc_status, qcdb_session, mlwh_session)
```
[the remark this answers is not in the document]
Comment (mgcam) [reply lang_qc/endpoints/pacbio_well.py:92]: Agree, but leaving this optimisation for future

File: lang_qc/models/pager.py [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+        title="Total number of unpaged objects",
+        description="""
+        Total number of unpaged objects. Needed by the UI in order to display
+        correct buttons in the paging widget.
+        """
+    )
+
+    @validator('total_number_of_items')
+    def check_total_number(cls, value):
+        if value < 0::
```
[the remark this answers is not in the document]
Comment (mgcam) [reply lang_qc/models/pager.py:71]: Agree, will remove the validator

File: lang_qc/endpoints/pacbio_well.py (hunk 1 above)
Comment (mgcam) [reply lang_qc/endpoints/pacbio_well.py:93]: Agree, should disappear in the next round of refactoring

File: lang_qc/endpoints/pacbio_well.py [hunk 4]
```diff
@@ -78,17 +78,26 @@
...
+    pbwells = pack_wells_and_states(wells, states)
+    paged_pbwells = pager.slice_data(pbwells)
+    return PacBioPagedWells(
+            total_number_of_items = len(pbwells),
+            data = pager.pselect_paged_data(paged_wells),
+            qc_flow_status = qc_status,
+            **pager
```
[the remark this answers is not in the document]
Comment (mgcam) [reply lang_qc/endpoints/pacbio_well.py:99]: This will be reworked.

File: lang_qc/db/utils.py [hunk 5]
```diff
@@ -33,12 +32,11 @@
...
 )
+from lang_qc.models.pacbio.well import WellStatusEnum
 
 
-def grab_recent_wells_from_db(
-    weeks: int, db_session: Session
-) -> List[PacBioRunWellMetrics]:
-    """Get wells from the past few weeks from the database."""
+def grab_wells_from_db(db_session: Session) -> List[PacBioRunWellMetrics]:
+    """Get completed wells from the from the database."""
```
Comment (nerdstrike) [inline lang_qc/db/utils.py:39]: Some repetition in this sentence

File: tests/test_filtered_wells.py [hunk 6]
```diff
@@ -35,14 +35,29 @@
...
 
+    response = test_client.get("/pacbio/wells")
+    assert response.status_code == 422  # page size and number should be given
+
+    response = test_client.get("/pacbio/wells?page_number=1")
+    assert response.status_code == 422  # page size should be given
 
-def assert_filtered_inbox_equals_expected(response, expected_data):
+    response = test_client.get("/pacbio/wells?page_size=1")
+    assert response.status_code == 422  # page number should be given
```
Comment (nerdstrike) [inline tests/test_filtered_wells.py:55]: After a fashion, these are testing pydantic and fastapi. Do we need these 422-type tests?

**Review by nerdstrike — CHANGES_REQUESTED** [review]
One typo, one question about test redundancy.

**Review by nerdstrike — APPROVED** [review]

**Pull request merged**